### PR TITLE
check for null fields when exporting Genome to Genbank, add unit tests for Genbank to Genome

### PR DESCRIPTION
**Pull request by jmchandonia**
This fixes the genbank exporter to check for null return values on optional fields in a Genome typed object.  Some of those fields are null for genomes created within KBase, which prevents them from exporting (https://atlassian.kbase.us/browse/KBASE-1721).  This should fix that issue, or at least it does for my test cases.

### CODE REVIEW

**Comment by jmchandonia** [comment]
This branch now adds some unit tests for Genbank->Genome conversion, and reverts a previous change by @realmarcin that caused the "develop" branch to fail some of the unit tests that the production code (in the master branch) passed.

**Comment by realmarcin** [comment]
Thanks for looking at this John-Marc. I haven't looked at your code edits yet in this request but I'm concerned that we have divergent test cases. The piece to add missing quotes was to address issues with upload of some V5 genomes. However, that test is too large for a unit test. We should revisit when the large tests gets rerun and perhaps identify more test cases for unit testing.

I also think that with these cases where the user-loaded file has what look like fixable issues we should issue a warning, have the user fix the problems, and then reuplod. Automatically fixing issues can get dangerous and the user a) might not be aware of the problem and b) what was done to address it.

**Comment by jmchandonia** [comment]
If you send me one of the large V5 genomes that required your code fix, I will test it and try to devise an appropriate unit test.  However, that code was never checked into the master branch or deployed in production, as far as I can tell.  It was causing unit tests based on perfectly standard format Genbank files to fail, which is why I reverted it.  Two such unit tests are based on Genbank files you checked in.

Also, I haven't checked in any code to check or handle the "nonstandard format" unit test cases yet. We first need to assay 1) what the current behavior is on such cases, and 2) what we want the behavior to be.  I'll argue for warnings rather than fatal errors in most cases, as well as no regressions that make the service reject files that it currently accepts.

**Comment by jmchandonia** [comment]
Tested and found these changes also fix https://atlassian.kbase.us/browse/KBASE-1524

**Comment by realmarcin** [comment]
OK, that sounds promising. This ENIGMA genome had multiple issues, most of
them pertaining to 'user-defined' fields not specified in the GenBank
format. Another issue was arbitrary dropping of quotes for non-enumerated
fields. I'm also not clear on the field names with this style
'/FIELD_NAME'. Also a number of these user-defined field names are acronyms
and should be capitalized (Pfam, KEGG, COG etc.). Finally, this user field
is particularly odd:

/ABC-type=cobalt
/ABC-type=transport
/ABC-type=system
/ABC-type=ATPase
/ABC-type=component

As it spans multiple lines, has an unclear label for the field (what is
'ABC-type'?), and has no quotes.

All of these user-defined fields are gene annotations and the GenBank
format allows to specify such annotations using the format specification.
Why not just use the GenBank sanctioned way? This will make it more likely
that KBase (and other tools) can load and use this additional data. Note
that these 'user-defined' fields also broke loading into genome viewers
like Artemis.

I know you've added code to skip over fields not specified explcitly in the
GenBank format, but I assume some warning is given?

**Comment by mlhenderson** [comment]
I see that you made some formatting changes to the original code, some of it I think is an improvement, but some is not.  I will line comment the code in specific areas where I have some issues.  We should establish some code styling guidelines to follow so that code remains consistent.

**Comment by MrCreosote** [comment]
Probably an APC type transporter.

-g

**Comment by realmarcin** [comment]
yep

http://en.wikipedia.org/wiki/Type_96_Armored_Personnel_Carrier

just unclear why it needs its own Genbank feature definition :)

**Comment by jmchandonia** [comment]
On Tue, Apr 14, 2015 at 12:05:01PM -0700, realmarcin wrote:
> As it spans multiple lines, has an unclear label for the field (what is
> 'ABC-type'?), and has no quotes.

Yes, there are other fields that also relate to gene annotations.
The Genbank file looks like it was generated with Biopython, which doesn't
appear to be too picky about its output format.

> I know you've added code to skip over fields not specified explcitly in the
> GenBank format, but I assume some warning is given?

Yes, it prints a warning on every such field.  Do "ant test" on my
branch to see examples.  I don't think those warnings get relayed by the
transform service to the user yet, but there needs to be some standard
for sending the user warnings (i.e., other than throwing a fatal exception).

Can you send me more examples of Genbank files that you needed custom
code to process?  I validated my code changes against all the files you'd
checked in to the t/genbank/ directory, but if there are other such cases
you know about, I can write more unit tests.

JMC
-- 
John-Marc Chandonia (JMChandonia@lbl.gov)               We're everywhere...
Physical Biosciences Division, Berkeley National Lab    for your convenience.
http://www.strgen.org/~jmc/                                 -- Psi Corps <*>

**Comment by realmarcin** [comment]
On Tue, Apr 14, 2015 at 1:19 PM, jmchandonia <notifications@github.com>
wrote:

> On Tue, Apr 14, 2015 at 12:05:01PM -0700, realmarcin wrote:
> > As it spans multiple lines, has an unclear label for the field (what is
> > 'ABC-type'?), and has no quotes.
>
> Yes, there are other fields that also relate to gene annotations.
> The Genbank file looks like it was generated with Biopython, which doesn't
> appear to be too picky about its output format.


I'm going to talk to Ryan Melnick about the Biopython part today ...


>
>
> > I know you've added code to skip over fields not specified explcitly in
> the
> > GenBank format, but I assume some warning is given?
>
> Yes, it prints a warning on every such field. Do "ant test" on my
> branch to see examples. I don't think those warnings get relayed by the
> transform service to the user yet, but there needs to be some standard
> for sending the user warnings (i.e., other than throwing a fatal
> exception).
>
> Can you send me more examples of Genbank files that you needed custom
> code to process? I validated my code changes against all the files you'd
> checked in to the t/genbank/ directory, but if there are other such cases
> you know about, I can write more unit tests.
>

 That data is from a monolithic test against all V5 genomes and at this
point I need to rerun that. I will let you know after Thur. this week.

File: src/us/kbase/genbank/GbkParser.java [hunk 1]
```diff
@@ -121,20 +125,10 @@ public static void parse(BufferedReader br, GbkParsingParams params, String file
...
-                                int chop = line.indexOf("/protein_id=");
-                                line = line.substring(0, chop);
-                            }
-                            if (!line.endsWith("\""))
-                                line += "\"";
-                            if (!line.equals("\"")) {
-                                qual = new GbkQualifier(line_num, qual_name, line);
-                                feat.qualifiers.add(qual);
-                            }
+			    line = line.substring(equal_pos+1).trim();
```
Comment (mlhenderson) [inline src/us/kbase/genbank/GbkParser.java]: Should keep the original spacing.

File: src/us/kbase/genbank/GbkParser.java [hunk 2]
```diff
@@ -121,20 +125,10 @@ public static void parse(BufferedReader br, GbkParsingParams params, String file
...
+			    if(qual!=null) qual.close();
```
Comment (mlhenderson) [inline src/us/kbase/genbank/GbkParser.java]: Should keep the original spacing.

File: src/us/kbase/genbank/GbkParser.java [hunk 3]
```diff
@@ -172,28 +166,28 @@ public static void main(String[] args) throws Exception {
         final String filename = "Pseudomonas_stutzeri_DSM_10701.gb";
         BufferedReader br = new BufferedReader(new FileReader(new File("test/" + filename)));
         parse(br, new GbkParsingParams(false), filename, new GbkCallback() {
-            @Override
-            public void setGenomeTrackFile(String contigName, String genomeName, int taxId, String plasmid, String filename) throws Exception {
-                pw.println("setGenome: contigName=" + contigName + ", genomeName=" + genomeName + ", taxId=" + taxId + ", plasmid=" + plasmid);
-            }
+		@Override
```
Comment (mlhenderson) [inline src/us/kbase/genbank/GbkParser.java]: Lines 169-190, what happened here?

File: src/us/kbase/genbank/GbkQualifier.java [hunk 4]
```diff
@@ -5,29 +5,32 @@
...
+    public GbkQualifier(int line_num,String type,String value) {
+	super(line_num,type,value);
+    }
+    public void close() {
+	is_closed = true;
+	String val = value.toString();
+	if (val.startsWith("\"")) {
+	    if (val.endsWith("\"")) {
+		is_text = true;
+		value.deleteCharAt(0);
```
Comment (mlhenderson) [inline src/us/kbase/genbank/GbkQualifier.java]: Future point to discuss, we should be compiling a list of these "corrections", what line numbers they represent, etc, and storing that in provenance.

File: src/us/kbase/genbank/GenometoGbk.java [hunk 5]
```diff
@@ -292,26 +292,28 @@ private StringBuffer createHeader(String contig_id, int contigssize, long contig
...
-            index++;
-            if (counter >= 65 || rawTaxonomy.length() < 80) {
-                formatTax.append("\n");
-                formatTax.append("            ");
-                counter = 0;
-            }
-        }
-
-        out.append("            " + formatTax + ".\n");
+	if (rawTaxonomy != null) {
```
Comment (mlhenderson) [inline src/us/kbase/genbank/GenometoGbk.java]: This chunk of code must have comments. Lines 295-316.

File: src/us/kbase/genbank/test/TestCommandLine.java [hunk 6]
```diff
@@ -25,7 +25,7 @@
  * Date: 10/29/14
  * Time: 12:17 AM
  */
-public class CommandLineTest {
+public class TestCommandLine {
```
Comment (mlhenderson) [inline src/us/kbase/genbank/test/TestCommandLine.java]: Is this a test driver/runner?  If we're renaming I would suggest TestDriverCommandLine.
I don't see specific tests in here to test the command line functionality itself, just something that can be used as a driver to run tests.

Comment (jmchandonia) [inline src/us/kbase/genbank/test/TestCommandLine.java]: I renamed @realmarcin's class so it wouldn't look like a class with JUnit unit tests.  Here
and elsewhere in KBase, those classes are in a test/ directory and also named *Test.java.
This seems to be the most popular convention:
http://stackoverflow.com/questions/3146821/naming-convention-junit-suffix-or-prefix-test
Besides FormatTest, which I added, I don't know what the other test classes do, or even if they contain any code that still tests anything useful.  But none of the others contain unit tests, and most seem to follow a Test*.java naming convention, so I renamed this one for consistency.

Comment (mlhenderson) [inline src/us/kbase/genbank/test/TestCommandLine.java]: That makes sense.  I'm going to give this particular one a pass for now,
but we do need to settle on some of these coding conventions generally and
have them be consistent across the project.

File: src/us/kbase/genbank/GbkQualifier.java (hunk 4 above)
Comment (mlhenderson) [inline src/us/kbase/genbank/GbkQualifier.java]: Also, a short code comment describing why this is here should exist.  I realize you are adding code here and the original was also not commented, but it needs some sort of description in the code.

File: src/us/kbase/genbank/GbkParser.java (hunk 2 above)
Comment (jmchandonia) [inline src/us/kbase/genbank/GbkParser.java]: This section was reverted from the non-working code in the "develop" branch to working code from the "master" branch.  At some point, somebody in the develop branch had changed the spacing, so my change restored the original.

File: src/us/kbase/genbank/GenometoGbk.java (hunk 5 above)
Comment (jmchandonia) [inline src/us/kbase/genbank/GenometoGbk.java]: Added a few comments here.  Marcin could provide further details.

File: src/us/kbase/genbank/GbkParser.java (hunk 2 above)
Comment (jmchandonia) [inline src/us/kbase/genbank/GbkParser.java]: I fixed both these lines manually.

File: src/us/kbase/genbank/GbkParser.java (hunk 3 above)
Comment (jmchandonia) [inline src/us/kbase/genbank/GbkParser.java]: Buggy version of Java emacs mode.  I patched it and re-indented everything; see commit below.

Comment (jmchandonia) [inline src/us/kbase/genbank/GbkParser.java]: Looks like a bug in the way emacs handles Java indentation.  Will patch & re-commit.

**Comment by mlhenderson** [comment]
Great, thanks!

**Pull request merged**